engine/src/main/java/org/pentaho/di/core/playlist/FilePlayListReplay.java

```diff
@@ -0,0 +1,113 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.core.playlist;
+
+import java.io.IOException;
+import java.util.Date;
+
+import org.apache.commons.vfs2.FileObject;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.trans.step.errorhandling.AbstractFileErrorHandler;
+
+public class FilePlayListReplay implements FilePlayList {
+
+  private final Date replayDate;
+
+  private final String encoding;
+
+  private final String lineNumberDirectory;
+
+  private final String lineNumberExtension;
+
+  private final String errorDirectory;
+
+  private final String errorExtension;
+
+  private FilePlayListReplayFile currentLineNumberFile;
+
+  private FilePlayListReplayFile currentErrorFile;
+
+  public FilePlayListReplay( Date replayDate, String lineNumberDirectory, String lineNumberExtension,
+    String errorDirectory, String errorExtension, String encoding ) {
+    this.replayDate = replayDate;
+    this.errorDirectory = errorDirectory;
+    this.errorExtension = errorExtension;
+    this.encoding = encoding;
+    this.lineNumberDirectory = lineNumberDirectory;
+    this.lineNumberExtension = lineNumberExtension;
+
+  }
+
+  private FileObject getCurrentProcessingFile() {
+    FileObject result = null;
+    if ( currentLineNumberFile != null ) {
+      result = currentLineNumberFile.getProcessingFile();
+    }
+    return result;
+  }
+
+  private String getCurrentProcessingFilePart() {
+    String result = null;
+    if ( currentLineNumberFile != null ) {
+      result = currentLineNumberFile.getProcessingFilePart();
+    }
+    return result;
+  }
+
+  public boolean isProcessingNeeded( FileObject file, long lineNr, String filePart ) throws KettleException {
+    initializeCurrentIfNeeded( file, filePart );
+    return currentLineNumberFile.isProcessingNeeded( file, lineNr, filePart )
+      || currentErrorFile.isProcessingNeeded( file, lineNr, filePart );
+  }
+
+  private void initializeCurrentIfNeeded( FileObject file, String filePart ) throws KettleException {
+    if ( !( file.equals( getCurrentProcessingFile() ) && filePart.equals( getCurrentProcessingFilePart() ) ) ) {
+      initializeCurrent( file, filePart );
+    }
+  }
+
+  private void initializeCurrent( FileObject file, String filePart ) throws KettleException {
+    try {
+      FileObject lineFile =
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *lineFile*. Currently, there are execution paths that do not contain closure statements, for example, when FileObject.exists throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).